##### Import dependencies and libraries

In [653]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

##### Load the csv file and display the first few rows of the data

In [654]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

##### Summary statistics

In [655]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

##### Plot the data to visualize market trends

In [656]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90,
    xlabel='coin_id',  # Label for the x-axis
    ylabel='Values',   # Label for the y-axis
    title='Market Data Trends'  # Add a title to the plot
)


:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [657]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
scaler = StandardScaler()
normalized_data = scaler.fit_transform(df_market_data)
normalized_data[:5]

array([[ 0.50852937,  0.49319307,  0.77220043,  0.23545963, -0.0674951 ,
        -0.35595348, -0.25163688],
       [ 0.18544589,  0.93444504,  0.55869212, -0.05434093, -0.27348273,
        -0.11575947, -0.19935211],
       [ 0.02177396, -0.70633685, -0.02168042, -0.06103015,  0.00800452,
        -0.55024692, -0.28206051],
       [-0.04076438, -0.81092807,  0.24945797, -0.05038797, -0.37316402,
        -0.45825882, -0.29554614],
       [ 1.19303608,  2.00095907,  1.76061001,  0.54584206, -0.29120287,
        -0.49984776, -0.27031695]])

##### Create a dataframe with scaled data

In [658]:
# Create a DataFrame with the scaled data
df_scaled_data = pd.DataFrame(normalized_data, columns=df_market_data.columns, index=df_market_data.index)

# Copy the crypto names from the original data
df_scaled_data['coin_id'] = df_market_data.index

# Set the coinid column as index
df_scaled_data.set_index('coin_id', inplace=True)

# Display sample data
df_scaled_data.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coin_id                                   
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

---

### Find the Best Value for k Using the Original Data.

In [659]:
# Create a list with the number of k-values from 1 to 11
k_values = list(range(1, 12))
k_values

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]

##### Compute inertia for each k-value

In [660]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
for k in k_values:
    # Inside the loop:
    # 1. Create a KMeans model using the loop counter for the n_clusters
    model = KMeans(n_clusters=k, random_state=0)
    # 2. Fit the model to the data using `df_market_data_scaled`
    model.fit(df_scaled_data)
    # 3. Append the model.inertia_ to the inertia list
    inertia.append(model.inertia_)

##### Create a dictionary with k-values and inertia
##### Convert the dictionary to a dataframe

In [661]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data = {
    'k': k_values,      # List of k-values
    'inertia': inertia  # Corresponding inertia values
}
# Create a DataFrame with the data to plot the Elbow curve
df_elbow = pd.DataFrame(elbow_data)

# Review the DataFrame
df_elbow.head()

k     inertia
0  1  287.000000
1  2  198.571818
2  3  145.635748
3  4   79.022435
4  5   66.217780

##### Plot the elbow curve

In [662]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
df_elbow.hvplot.line(
    x='k', 
    y='inertia', 
    title='Elbow Curve for Optimal k (Original Data)',
    xticks=k, 
    xlabel='Number of Clusters (k)', 
    ylabel='Inertia', 
    width=800, 
    height=400
)

:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** Based on the Elbow Curve, the best value for k is 4.

- The "elbow point" is where the inertia starts to decrease at a slower rate, which indicates diminishing returns from adding more clusters.
- In this case, the slope of the curve flattens significantly after k=4, suggesting that 4 is the optimal number of clusters for this data.

---

### Cluster Cryptocurrencies with K-means Using the Original Data

##### Initialize kmeans model

In [663]:
# Initialize the K-Means model using the best value for k
k = 4  # Based on the Elbow method
kmeans_model = KMeans(n_clusters=k, random_state=0)

##### Fit the KMeans model to the scaled data to learn the clusters.

In [664]:
# Fit the K-Means model using the scaled data
kmeans_model.fit(df_scaled_data)

KMeans(n_clusters=4, random_state=0)

##### Predict and print the cluster labels for each cryptocurrency using the KMeans model.

In [665]:
# Predict the clusters to group the cryptocurrencies using the scaled data
cluster_predictions = kmeans_model.predict(df_scaled_data)

# Print the resulting array of cluster values.
print("Cluster labels for each cryptocurrency:")
cluster_predictions

Cluster labels for each cryptocurrency:


array([2, 2, 0, 0, 2, 2, 2, 2, 2, 0, 0, 0, 0, 2, 0, 2, 0, 0, 2, 0, 0, 2,
       0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 1, 2, 0, 0, 3, 0, 0, 0, 0],
      dtype=int32)

##### Copy of original dataframe

In [666]:
# Create a copy of the DataFrame
df_crypto_clusters = df_market_data.copy()

##### Added new column with predicted clusters

In [667]:
# Add a new column to the DataFrame with the predicted clusters
df_crypto_clusters['Cluster'] = cluster_predictions

# Display sample data
df_crypto_clusters.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                           1.08388                     7.60278   
ethereum                          0.22392                    10.38134   
tether                           -0.21173                     0.04935   
ripple                           -0.37819                    -0.60926   
bitcoin-cash                      2.90585                    17.09717   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                           6.57509                      7.67258   
ethereum                          4.80849                      0.13169   
tether                            0.00640                     -0.04237   
ripple                            2.24984                      0.23455   
bitcoin-cash                     14.75334                     15.74903   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                          -3.25185                      83.51840   
ethereum                        -12.88890                     186.77418   
tether                            0.28037                      -0.00542   
ripple                          -17.55245                      39.53888   
bitcoin-cash                    -13.71793                      21.66042   

              price_change_percentage_1y  Cluster  
coin_id                                            
bitcoin                         37.51761        2  
ethereum                       101.96023        2  
tether                           0.01954        0  
ripple                         -16.60193        0  
bitcoin-cash                    14.49384        2

##### Scatter plot of cryptocurrency clusters based on 24h vs 7d price changes, colored by KMeans cluster labels with hover information showing crypto names and price changes.

In [668]:
# Create a scatter plot using hvPlot by setting 
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`.
scatter_plot = df_crypto_clusters.hvplot.scatter(
    x='price_change_percentage_24h',         # X-axis feature
    y='price_change_percentage_7d',          # Y-axis feature
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.          
    c='Cluster',                             
    hover_cols=['coin_id'],
    cmap='Category10',                       # Use a darker, more distinct color map                  
    title='Cryptocurrency Clusters (Original data)',         # Add a title
    xlabel='24h Price Change (%)',           # Label for the X-axis
    ylabel='7d Price Change (%)',            # Label for the Y-axis
    width=800,                               # Plot width
    height=600                               # Plot height
)

# Display the scatter plot
scatter_plot

:Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,Cluster,coin_id)

---

### Optimize Clusters with Principal Component Analysis.

##### Create a PCA model to reduce the data to 3 principal components.

In [669]:
# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components=3)

##### Reduce the data to 3 principal components and show the first 5 rows.

In [670]:
# Use the PCA model with `fit_transform` to reduce to 
# three principal components.
pca_data = pca.fit_transform(df_scaled_data)

# View the first five rows of the DataFrame.
pca_data[:5]

array([[-0.60066733,  0.84276006,  0.46159457],
       [-0.45826071,  0.45846566,  0.95287678],
       [-0.43306981, -0.16812638, -0.64175193],
       [-0.47183495, -0.22266008, -0.47905316],
       [-1.15779997,  2.04120919,  1.85971527]])

##### Print the explained variance for each principal component and the total.

In [671]:
# Retrieve the explained variance to determine how much information 
# can be attributed to each principal component.
explained_variance = pca.explained_variance_ratio_

# Print the explained variance ratio for each of the three principal components
print("Explained Variance Ratio for each principal component:")
print(f"PC1: {explained_variance[0]:.4f}")
print(f"PC2: {explained_variance[1]:.4f}")
print(f"PC3: {explained_variance[2]:.4f}")

# Calculate and print the total explained variance of the three components
total_explained_variance = explained_variance.sum()
print(f"Total Explained Variance (PC1 + PC2 + PC3): {total_explained_variance:.4f}")

Explained Variance Ratio for each principal component:
PC1: 0.3720
PC2: 0.3470
PC3: 0.1760
Total Explained Variance (PC1 + PC2 + PC3): 0.8950


#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** The total explained variance of the three principal components is 0.8950, which means that these three components together explain 89.50% of the variability in the dataset.

##### Create PCA DataFrame, set coin_id as index, and display data

In [672]:
# Create a new DataFrame with the PCA data.
# Creating a DataFrame with the PCA data
df_pca = pd.DataFrame(pca_data, columns=['PC1', 'PC2', 'PC3'])

# Copy the crypto names from the original data
df_pca['coin_id'] = df_market_data.index

# Set the coinid column as index
df_pca.set_index('coin_id', inplace=True)

# Display sample data
df_pca.head()

PC1       PC2       PC3
coin_id                                   
bitcoin      -0.600667  0.842760  0.461595
ethereum     -0.458261  0.458466  0.952877
tether       -0.433070 -0.168126 -0.641752
ripple       -0.471835 -0.222660 -0.479053
bitcoin-cash -1.157800  2.041209  1.859715

---

### Find the Best Value for k Using the PCA Data

##### List of k-values for clustering

In [673]:
# Create a list with the number of k-values from 1 to 11
k_values = list(range(1, 12))

##### Compute and store inertia for each k using PCA data.

In [674]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
for k in k_values:
    # Inside the loop:
    # 1. Create a KMeans model using the loop counter for the n_clusters
    model = KMeans(n_clusters=k, random_state=0)
    # 2. Fit the model to the data using `df_market_data_pca`
    model.fit(df_pca)
    # 3. Append the model.inertia_ to the inertia list
    inertia.append(model.inertia_)


##### Create a dictionary and DataFrame to plot the Elbow curve.

In [675]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data = {
    'k': k_values,      # List of k-values
    'inertia': inertia  # Corresponding inertia values
}
# Create a DataFrame with the data to plot the Elbow curve
df_elbow_pca = pd.DataFrame(elbow_data)

# Display the first few rows of the DataFrame
df_elbow_pca.head()


k     inertia
0  1  256.874086
1  2  168.811896
2  3  115.918077
3  4   49.665497
4  5   41.357314

##### Plot the Elbow curve to visually find the optimal k

In [676]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
df_elbow_pca.hvplot.line(
    x='k', 
    y='inertia', 
    title='Elbow Curve for Optimal k (PCA)',
    xticks=k_values, 
    xlabel='Number of Clusters (k)', 
    ylabel='Inertia', 
    width=800, 
    height=400
)


:Curve   [k]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** The best value for k appears to be 4, as this is where the curve begins to flatten, indicating diminishing returns in reducing the inertia with additional clusters.


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** No, it does not differ from the best k value found using the original data, which was also 4.

### Cluster Cryptocurrencies with K-means Using the PCA Data

##### Initialize the KMeans model with k=4

In [677]:
# Initialize the K-Means model using the best value for k
k = 4  # Based on the Elbow method
kmeans_pca = KMeans(n_clusters=k, random_state=0)


##### Fit the KMeans model to the PCA data

In [678]:
# Fit the K-Means model using the PCA data
kmeans_pca.fit(df_pca)

KMeans(n_clusters=4, random_state=0)

##### Predict cluster labels for the PCA data.

In [679]:
# Predict the clusters to group the cryptocurrencies using the PCA data
pca_cluster_predictions = kmeans_pca.predict(df_pca)

# Print the resulting array of cluster values.
print("Cluster labels for each cryptocurrency using PCA data:")
pca_cluster_predictions


Cluster labels for each cryptocurrency using PCA data:


array([2, 2, 0, 0, 2, 2, 2, 2, 2, 0, 0, 0, 0, 2, 0, 2, 0, 0, 2, 0, 0, 2,
       0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 1, 2, 0, 0, 3, 0, 0, 0, 0],
      dtype=int32)

##### Create a copy of the PCA dataframe and add a column with predicted clusters.

In [680]:
# Create a copy of the DataFrame with the PCA data
df_pca_clusters = df_pca.copy()

# Add a new column to the DataFrame with the predicted clusters
df_pca_clusters['Cluster'] = pca_cluster_predictions

# Display sample data
df_pca_clusters.head()


PC1       PC2       PC3  Cluster
coin_id                                            
bitcoin      -0.600667  0.842760  0.461595        2
ethereum     -0.458261  0.458466  0.952877        2
tether       -0.433070 -0.168126 -0.641752        0
ripple       -0.471835 -0.222660 -0.479053        0
bitcoin-cash -1.157800  2.041209  1.859715        2

##### Create the scatter plot with PCA components and clusters.

In [681]:
# Create a scatter plot using hvPlot by setting 
# `x="PC1"` and `y="PC2"`.
scatter_plot_pca = df_pca_clusters.hvplot.scatter(
    x='PC1',                                 # X-axis feature: First principal component
    y='PC2',                                 # Y-axis feature: Second principal component
    c='Cluster',                             # Color the graph points with the labels found using K-Means
    hover_cols=['coin_id'],                  # Add the crypto name in the `hover_cols` parameter to identify the cryptocurrency represented by each data point
    cmap='Category10',                       # Use a darker, more distinct color map
    title='Cryptocurrency Clusters (PCA)',   # Add a title
    xlabel='Principal Component 1 (PC1)',    # Label for the X-axis
    ylabel='Principal Component 2 (PC2)',    # Label for the Y-axis
    width=800,                               # Plot width
    height=600                               # Plot height
)
# Display the scatter plot 
scatter_plot_pca


:Scatter   [PC1]   (PC2,Cluster,coin_id)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

##### Create a composite plot to compare the Elbow curves for the original data and the PCA data

In [682]:
# Composite plot to contrast the Elbow curves
# Elbow curve for the original data
elbow_plot_original = df_elbow.hvplot.line(
    x='k', 
    y='inertia', 
    #title='Elbow Curve Comparison (Original vs. PCA)',
    xticks=k_values, 
    xlabel='Number of Clusters (k)', 
    ylabel='Inertia', 
    width=400, 
    height=400,
    color='blue',
    title ='Elbow Curve (Without PCA)'
) 

# Elbow curve for the PCA data
elbow_plot_pca = df_elbow_pca.hvplot.line(
    x='k', 
    y='inertia', 
    xticks=k_values, 
    xlabel='Number of Clusters (k)', 
    ylabel='Inertia', 
    width=400, 
    height=400,
    color='red',
    title ='Elbow Curve(With PCA)'
)

# Composite plot: Overlay both Elbow curves (original vs PCA)
composite_elbow_plot = elbow_plot_original + elbow_plot_pca

# Display the plot
composite_elbow_plot

:Layout
   .Curve.I  :Curve   [k]   (inertia)
   .Curve.II :Curve   [k]   (inertia)

##### Create a composite plot to compare the scatter plots for the original data and the PCA data

In [683]:
# Scatter plot for clusters from original data 
scatter_plot_original = df_crypto_clusters.hvplot.scatter(
    x='price_change_percentage_24h',         # X-axis feature for the original data
    y='price_change_percentage_7d',          # Y-axis feature for the original data
    c='Cluster',                             # Color points by cluster labels
    hover_cols=['coin_id'],                  # Show cryptocurrency name on hover
    cmap='Category10',                       # Use a categorical color map
    title='Clusters from Original Data',     # Title for the original data plot
    xlabel='24h Price Change (%)',           # X-axis label
    ylabel='7d Price Change (%)',            # Y-axis label
    width=400,                               # Plot width
    height=400                               # Plot height
)

# Scatter plot for clusters from PCA-transformed data
scatter_plot_pca = df_pca_clusters.hvplot.scatter(
    x='PC1',                                 # X-axis: First principal component
    y='PC2',                                 # Y-axis: Second principal component
    c='Cluster',                             # Color points by cluster labels
    hover_cols=['coin_id'],                  # Show cryptocurrency name on hover
    cmap='Category10',                       # Use a categorical color map
    title='Clusters from PCA Data',          # Title for the PCA-transformed data plot
    xlabel='Principal Component 1',          # X-axis label
    ylabel='Principal Component 2',          # Y-axis label
    width=400,                               # Plot width
    height=400                               # Plot height
)

# Use the + operator to display both scatter plots side by side
composite_cluster_plot = scatter_plot_original + scatter_plot_pca

# Display the composite plot
composite_cluster_plot

:Layout
   .Scatter.I  :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,Cluster,coin_id)
   .Scatter.II :Scatter   [PC1]   (PC2,Cluster,coin_id)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** 

Based on the visual comparison: 
##### 1. Elbow Curves:
Both the Elbow curves with and without PCA show a similar optimal k value, 𝑘 = 4.
However, the PCA-transformed data has a slightly lower inertia across all values of k. This indicates that PCA reduces the overall variance, likely by removing irrelevant features or noise, leading to more efficient clustering with fewer features.
##### 2. Cluster Comparison:
In the original data, the clusters are spread across two distinct features (24h and 7d price changes). While clusters are identifiable, some overlap exists, and data points might not be as cleanly separated.
In the PCA-transformed data, the clusters seem more compact and better separated, especially in the lower-dimensional space (PC1 vs PC2). This suggests that reducing dimensionality with PCA improves the separation of clusters, making them more distinct and easier to interpret.
##### Conclusion:
Using fewer features through PCA simplifies the clustering process, potentially improving cluster separation and reducing noise. The reduction in dimensionality doesn't appear to sacrifice significant information since the clustering results are similar, but with clearer boundaries in the PCA-transformed data. However, the downside is the interpretability of the principal components compared to the original features.